pynamodb/tests/integration/test_transaction_integration.py

```diff
@@ -0,0 +1,218 @@
+
+    @classmethod
+    def setup_class(cls):
+        # ensure that the transaction connection url is what we expect
+        Transaction(host=DDB_URL, override_connection=True)
+        for m in TEST_MODELS:
+            if not m.exists():
+                m.create_table(
+                    read_capacity_units=10,
+                    write_capacity_units=10,
+                    wait=True
+                )
+
+    @classmethod
+    def teardown_class(cls):
+        for m in TEST_MODELS:
+            if m.exists():
+                m.delete_table()
+
+    @staticmethod
+    def get_error_code(error):
+        return error.cause.response['Error'].get('Code')
+
+    @pytest.mark.ddblocal
+    def test_transact_write__error__idempotent_parameter_mismatch(self):
+
+        transaction = TransactWrite(host=DDB_URL)
+        User(1).save(in_transaction=transaction)
+        User(2).save(in_transaction=transaction)
+
+        try:
+            # committing the first time, then adding more info and committing again
+            transaction.commit()
+            User(3).save(in_transaction=transaction)
+            transaction.commit()
+            assert False, 'Failed to raise error'
+        except PutError as e:
+            assert self.get_error_code(e) == IDEMPOTENT_PARAMETER_MISMATCH
+
+    @pytest.mark.ddblocal
+    def test_transact_write__error__different_regions(self):
+        # creating a model in a table outside the region everyone else operates in
+        transact_write = TransactWrite()
+        DifferentRegion(entry_index=0).save(in_transaction=transact_write)
+        BankStatement(1).save(in_transaction=transact_write)
+        User(1).save(in_transaction=transact_write)
+
+        try:
+            transact_write.commit()
+        except PutError as e:
+            assert self.get_error_code(e) == RESOURCE_NOT_FOUND
+
+    @pytest.mark.ddblocal
+    def test_transact_write__error__transaction_cancelled(self):
+        # create a users and a bank statements for them
+        User(1).save()
+        BankStatement(1).save()
+
+        # attempt to do this as a transaction with the condition that they don't already exist
+        transaction = TransactWrite()
+        User(1).save(condition=(User.user_id.does_not_exist()), in_transaction=transaction)
+        BankStatement(1).save(condition=(BankStatement.user_id.does_not_exist()), in_transaction=transaction)
+        try:
+            transaction.commit()
+            assert False, 'Failed to raise error'
+        except PutError as e:
+            assert self.get_error_code(e) == TRANSACTION_CANCELLED
+
+    @pytest.mark.ddblocal
+    def test_transact_get(self):
+        # making sure these entries exist, and with the expected info
+        User(1).save()
+        BankStatement(1).save()
+        User(2).save()
+        BankStatement(2, balance=100).save()
+
+        # get users and statements we just created and assign them to variables
+        with TransactGet() as transaction:
```

Comment:
In general I like this interface, though I'm worried that either:
 1) People will accidentally try to use properties of these models before the transaction is committed
 2) A single item will hit some error condition [1] and then the proxy object will be invalid

Did you consider returning some _other_ object which the hydrated model can be fetched? That other object would raise exceptions if things are not right (the transaction is still open, the item doesn't exist, etc.)

cc @ikonst for opinions


[1]  https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_TransactGetItems.html
> If a requested item could not be retrieved, the corresponding ItemResponse object is Null, or if the requested item has no projected attributes, the corresponding ItemResponse object is an empty Map.

Reply:
Do you mean like making a class that inherits from `Model` that will raise an error if some attribute (`transaction_in_progress = True`?) indicates that it hasn't completed and the user attempts to access any properties before the commit? Then updates the model with all of the attributes at the end?

I guess my general worry would be how do we make it clear to the user that the models returned by a transaction aren't the same model classes that they'd have in a normal `get`

Reply:
Something like:
```py
with TransactGet() as transaction:
    user1_promise = User.get(1, in_transaction=transaction)

user1 = user1_promise.resolve()
```
?

I suppose this could be typed as:
```
_T = TypeVar('_T', bound='Model')

class _ModelPromise[_T]:
    def resolve() -> _T: ...

class Model(metaclass=MetaModel):
    ...
    @override
    @classmethod
    def get(cls: Type[_T], hash_key: KeyType, range_key: Optional[KeyType] = ..., consistent_read: bool = ..., in_transaction: Transaction) -> _ModelPromise[_T]: ...

    @override
    @classmethod
    def get(cls: Type[_T], hash_key: KeyType, range_key: Optional[KeyType] = ..., consistent_read: bool = ...) -> _T: ...
```

Reply:
I guess the primary downside to this is that you can't do `user1: User = User.get(...)`, and if you wanted `User` typing on the variable, you'd have to reassign it outside of the transaction like:
```
with TransactWrite() as t:
    _user1_promise = User.get(1, in_transaction=t)
    _user2_promise = User.get(2, in_transaction=t)
user1: User = _user1_promise.resolve()
user2: User = _user2_promise.resolve()
```

Reply:
To clarify, you don't need the `: User` part - mypy should be smart enough.

Reply:
@jpinner-lyft says he'd rather
```py
with TransactWrite() as t:
    user1_future = t.get(User, 1)
    ...
user1 = user1_future.get()
```

(a) let's call it a "future" rather than a "promise" as per:
> A future is a placeholder object for a result that does not yet exist. A promise is a writable, single-assignment container, which completes a future. Promises can complete the future with a result to indicate success, or with an exception to indicate failure.
> ...
> Promises can be thought of as a future with a public set method which the caller (or anybody else) can use to set the value of the future.

Since in our case the caller (i.e. user code) cannot set the value of the future, then the caller is merely given a "future".

(b) how about we rename "resolve" to "get" as per
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html#get()
?

Reply:
Yep, I'm on board with the future idea. I think we should move back to having the `.get` and `.delete`, etc. as methods on the `Transaction` object rather than overloading the ones on `Model`. 

e.g
```python
with TransactGet() as t:
  user_future = t.get(User, hash_key, range_key)

try:
  user = user_future.get()
except User.DoesNotExist:
  pass
except ...
```

Reply:
Let's `resolve()` —> `get()`, or maybe `value` (property). I've made the mistake of suggesting `resolve()` but it's actually a method of a _promise_ which should get the value as an argument (i.e. `promise.resolve(to_value)`).

Reply:
@ikonst Yep, you're right there. I haven't worked with promise semantics for a while now so forgot some of the specifics :)

Reply:
So, to be clear on the ask, we want `TransactGet` to have a completely different interface from `TransactWrite`, in that, instead of overloading the `Model.get` function, we create a `TransactGet.get`, that works like:
```python
with TransactGet() as t:
    user_future: _ModelFuture = t.get(model=User, hash_key=1)
    ride_future: _ModelFuture = t.get(model=Ride, hash_key=1, range_key=2)

try:
    user = user_future.get()
    ride = ride_future.get()
except DoesNotExist:
    pass
```

?

Reply:
> instead of overloading the Model.get function, we create a TransactGet.get, that works like ...
Yep, that's the suggestion.

> we want TransactGet to have a completely different interface from TransactWrite
Nope! I would suggest that we adopt a similar interface with `TransactWrite` for consistency. Happy to talk through what that should look like